File: mikkel_agent/opponent_model.py
from geniusweb.issuevalue.Bid import Bid
from geniusweb.profileconnection.ProfileInterface import ProfileInterface

from agents.mikkel_agent.boa_element import BoaElement
from agents.mikkel_agent.helpers.negotiation_progress import NegotiationProgress


class OpponentModel(BoaElement):

    _proposed_bids: [(Bid, float)] = []

    def __init__(self, progress: NegotiationProgress, profile: ProfileInterface):
        super().__init__(progress, profile)

    def record_bid(self, bid: Bid):
        self._proposed_bids.append((bid, self.get_utility(bid)))

    def peek_best_proposed_utility(self) -> float:
        return self._get_best_proposed_bid()[1]

    def get_best_proposed_bid(self) -> Bid:
        best_entry = self._get_best_proposed_bid()
        self._proposed_bids.remove(best_entry)
        return best_entry[0]

    def _get_best_proposed_bid(self) -> (Bid, float):
        return max(self._proposed_bids, key=lambda proposal: proposal[1])

File: areaparty/src/main/java/geniusweb/exampleparties/areaparty/OpponentModel.java
package geniusweb.exampleparties.areaparty;

import geniusweb.issuevalue.Bid;
import lombok.experimental.SuperBuilder;

import java.util.Comparator;
import java.util.LinkedList;
import java.util.List;
import java.util.Optional;


@SuperBuilder
public class OpponentModel extends BoaElement {

    private final List<EvaluatedBid> proposedBids = new LinkedList<>();

    public void recordBid(Bid bid) {
        this.proposedBids.add(new EvaluatedBid(bid, this.getUtility(bid)));
    }

    public Bid getBestProposedBid() {
        return this.peekBestProposedBid().getBid();
    }

    public double peekBestProposedUtility() {
        return this.peekBestProposedBid().getUtility();
    }

    private EvaluatedBid peekBestProposedBid() {
        Optional<EvaluatedBid> bestBid = this.proposedBids.stream()
                .max(Comparator.comparingDouble(EvaluatedBid::getUtility));
        return bestBid.orElseThrow(() -> new RuntimeException("No best bid"));
    }
}
